File: riveredge-frontend/src/apps/kuaiplm/services/annual-lab-plan.ts
/**
 * 年度实验计划 API（R-07）
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/annual-lab-plans';

export type AnnualLabPlanStatus = 'draft' | 'pending' | 'approved' | 'rejected' | 'closed';
export type AnnualLabMonthStatus = 'pending' | 'in_progress' | 'completed';
export type AnnualLabIssueStatus =
  | 'none'
  | 'pending_dept'
  | 'pending_sales'
  | 'pending_plan'
  | 'approved'
  | 'rejected';

export interface AnnualLabPlanMonth {
  id?: number;
  uuid?: string;
  plan_id?: number;
  year_month?: string;
  month_no?: number;
  title?: string | null;
  month_status?: AnnualLabMonthStatus | string;
  owner_user_id?: number | null;
  owner_user_name?: string | null;
  due_at?: string | null;
  material_desc?: string | null;
  issue_status?: AnnualLabIssueStatus | string;
  issue_submitted_at?: string | null;
  issue_approved_at?: string | null;
  issue_reject_reason?: string | null;
  lab_request_id?: number | null;
  lab_request_code?: string | null;
  report_file_uuid?: string | null;
  report_url?: string | null;
  defect_desc?: string | null;
  treatment_result?: string | null;
  completed_at?: string | null;
  remarks?: string | null;
}

export interface AnnualLabPlan {
  id?: number;
  uuid?: string;
  plan_code?: string;
  plan_year?: number;
  title?: string;
  status?: AnnualLabPlanStatus | string;
  plan_file_uuid?: string | null;
  plan_file_name?: string | null;
  owner_user_id?: number | null;
  owner_user_name?: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  approved_by_name?: string | null;
  rejected_at?: string | null;
  reject_reason?: string | null;
  closed_at?: string | null;
  remarks?: string | null;
  months?: AnnualLabPlanMonth[];
  completed_months?: number;
  total_months?: number;
  created_by_name?: string | null;
  updated_by_name?: string | null;
  created_at?: string;
  updated_at?: string;
}

function unwrapList(res: unknown): { items: AnnualLabPlan[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as AnnualLabPlan[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const annualLabPlanApi = {
  list: async (params?: Record<string, unknown>) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => (await api.get(`${BASE}/${id}`)) as AnnualLabPlan,
  create: async (data: Partial<AnnualLabPlan>) =>
    (await api.post(BASE, data)) as AnnualLabPlan,
  update: async (id: number, data: Partial<AnnualLabPlan>) =>
    (await api.put(`${BASE}/${id}`, data)) as AnnualLabPlan,
  delete: async (id: number) => api.delete(`${BASE}/${id}`),
  submit: async (id: number) => (await api.post(`${BASE}/${id}/submit`)) as AnnualLabPlan,
  approve: async (id: number) => (await api.post(`${BASE}/${id}/approve`)) as AnnualLabPlan,
  reject: async (id: number, reason: string) =>
    (await api.post(`${BASE}/${id}/reject`, { reason })) as AnnualLabPlan,
  close: async (id: number) => (await api.post(`${BASE}/${id}/close`)) as AnnualLabPlan,
  updateMonth: async (planId: number, monthId: number, data: Partial<AnnualLabPlanMonth>) =>
    (await api.put(`${BASE}/${planId}/months/${monthId}`, data)) as AnnualLabPlan,
  submitIssue: async (planId: number, monthId: number) =>
    (await api.post(`${BASE}/${planId}/months/${monthId}/issue/submit`)) as AnnualLabPlan,
  approveIssue: async (planId: number, monthId: number) =>
    (await api.post(`${BASE}/${planId}/months/${monthId}/issue/approve`)) as AnnualLabPlan,
  rejectIssue: async (planId: number, monthId: number, reason: string) =>
    (await api.post(`${BASE}/${planId}/months/${monthId}/issue/reject`, { reason })) as AnnualLabPlan,
};

File: riveredge-backend/src/apps/kuaiplm/schemas/annual_lab_plan.py
"""年度实验计划 Schema（R-07）。"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class AnnualLabPlanMonthBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    year_month: str = Field(..., max_length=7)
    month_no: int = Field(..., ge=1, le=12)
    title: Optional[str] = Field(None, max_length=200)
    month_status: str = "pending"
    owner_user_id: Optional[int] = None
    owner_user_name: Optional[str] = Field(None, max_length=100)
    due_at: Optional[datetime] = None
    material_desc: Optional[str] = None
    issue_status: str = "none"
    lab_request_id: Optional[int] = None
    lab_request_code: Optional[str] = Field(None, max_length=50)
    report_file_uuid: Optional[str] = Field(None, max_length=36)
    report_url: Optional[str] = Field(None, max_length=500)
    defect_desc: Optional[str] = None
    treatment_result: Optional[str] = None
    remarks: Optional[str] = None


class AnnualLabPlanMonthInput(BaseModel):
    year_month: Optional[str] = Field(None, max_length=7)
    month_no: Optional[int] = Field(None, ge=1, le=12)
    title: Optional[str] = Field(None, max_length=200)
    owner_user_id: Optional[int] = None
    owner_user_name: Optional[str] = Field(None, max_length=100)
    due_at: Optional[datetime] = None
    material_desc: Optional[str] = None
    remarks: Optional[str] = None


class AnnualLabPlanMonthUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=200)
    month_status: Optional[str] = Field(None, max_length=30)
    owner_user_id: Optional[int] = None
    owner_user_name: Optional[str] = Field(None, max_length=100)
    due_at: Optional[datetime] = None
    material_desc: Optional[str] = None
    lab_request_id: Optional[int] = None
    lab_request_code: Optional[str] = Field(None, max_length=50)
    report_file_uuid: Optional[str] = Field(None, max_length=36)
    report_url: Optional[str] = Field(None, max_length=500)
    defect_desc: Optional[str] = None
    treatment_result: Optional[str] = None
    remarks: Optional[str] = None


class AnnualLabPlanMonthResponse(AnnualLabPlanMonthBase):
    id: int
    uuid: str
    plan_id: int
    issue_submitted_at: Optional[datetime] = None
    issue_approved_at: Optional[datetime] = None
    issue_reject_reason: Optional[str] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class AnnualLabPlanCreate(BaseModel):
    plan_code: Optional[str] = Field(None, max_length=50)
    plan_year: int = Field(..., ge=2000, le=2100)
    title: str = Field(..., max_length=200)
    plan_file_uuid: Optional[str] = Field(None, max_length=36)
    plan_file_name: Optional[str] = Field(None, max_length=255)
    owner_user_id: Optional[int] = None
    owner_user_name: Optional[str] = Field(None, max_length=100)
    remarks: Optional[str] = None
    months: Optional[List[AnnualLabPlanMonthInput]] = None


class AnnualLabPlanUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=200)
    plan_file_uuid: Optional[str] = Field(None, max_length=36)
    plan_file_name: Optional[str] = Field(None, max_length=255)
    owner_user_id: Optional[int] = None
    owner_user_name: Optional[str] = Field(None, max_length=100)
    remarks: Optional[str] = None
    months: Optional[List[AnnualLabPlanMonthInput]] = None


class AnnualLabPlanRejectRequest(BaseModel):
    reason: str = Field(..., min_length=1)


class AnnualLabPlanIssueRejectRequest(BaseModel):
    reason: str = Field(..., min_length=1)


class AnnualLabPlanResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: str
    plan_code: str
    plan_year: int
    title: str
    status: str
    plan_file_uuid: Optional[str] = None
    plan_file_name: Optional[str] = None
    owner_user_id: Optional[int] = None
    owner_user_name: Optional[str] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    approved_by: Optional[int] = None
    approved_by_name: Optional[str] = None
    rejected_at: Optional[datetime] = None
    reject_reason: Optional[str] = None
    closed_at: Optional[datetime] = None
    remarks: Optional[str] = None
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    months: List[AnnualLabPlanMonthResponse] = Field(default_factory=list)
    completed_months: int = 0
    total_months: int = 0


class AnnualLabPlanListResponse(BaseModel):
    data: List[AnnualLabPlanResponse] = Field(default_factory=list)
    total: int = 0
    success: bool = True
